connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginDesc.java

```diff
@@ -115,6 +115,10 @@ public int compareTo(PluginDesc<T> other) {
         int versionComp = encodedVersion.compareTo(other.encodedVersion);
         // isolated plugins appear after classpath plugins when they have identical versions.
         int isolatedComp = Boolean.compare(other.loader instanceof PluginClassLoader, loader instanceof PluginClassLoader);
-        return nameComp != 0 ? nameComp : (versionComp != 0 ? versionComp : isolatedComp);
+        // choose an arbitrary order between different classloaders and types
+        int loaderComp = loader.hashCode() - other.loader.hashCode();
```

Comment:
Perhaps if we want to reduce the likelihood of collisions, we can use `System::identityHashCode`?
```suggestion
        int loaderComp = System.identityHashCode(loader) - System.identityHashCode(other.loader)
```

Reply:
As far as I can tell, the PluginClassLoader is already using the identityHashCode internally, so I don't think this makes collisions any more or less likely.

Reply:
I've replaced this with a comparison of the locations, and tweaked some signatures to try to enforce that Paths (and thus locations) are unique.